bolt-modules/boltlib/lib/puppet/functions/resource.rb

```diff
@@ -0,0 +1,46 @@
+# frozen_string_literal: true
+
+# Lookup a resource in the target's data.
+#
+# For more information about resources see [the
+# documentation](https://puppet.com/docs/puppet/latest/lang_resources.html).
+#
+# > **Note:** The `ResourceInstance` data type is under active development and is subject to
+#   change. You can read more about the data type in the [experimental features
+#   documentation](experimental_features.md#resourceinstance-data-type).
+#
+# > **Note:** Not available in apply block
+Puppet::Functions.create_function(:resource) do
+  # Lookup a resource in the target's data.
+  # @param target The `Target` object to add resources to. See {get_targets}.
+  # @param type The type of the resource
+  # @param title The title of the resource
+  # @return The ResourceInstance if found, or Undef
+  # @example Get the openssl package resource
+  #   $target.apply_prep
+  #   $resources = $target.get_resources(Package).first['resources']
+  #   $target.set_resources($resources)
+  #   $openssl = $target.resource('Package', 'openssl')
+  dispatch :resource do
+    param 'Boltlib::TargetSpec', :target
+    param 'Variant[String[1], Type]', :type
+    param 'String[1]', :title
+    return_type 'Optional[ResourceInstance]'
+  end
+
+  def resource(target, type, title)
+    unless Puppet[:tasks]
```

Comment:
```suggestion
    unless Puppet[:tasks]
```

It seems a little odd that you can use `$target.resources` in an apply, but not `$target.resource`.

Reply:
In order to do that we'd need to make this a function on the target datatype itself. We can do that, but then we won't get analytics for it and `resource('target', 'type', 'title')` won't work. I'm not opposed to having the function be on the target - it's a small thing, but I didn't want to have to add a new "parameters" column to the datatype reference table just for this one function, and it seemed to make more sense as a plan function anyway.

Reply:
I'm not clear why this wouldn't work. What about an apply block requires it to be a function on the Target datatype?

Reply:
I guess 'wouldn't work' is the wrong way of phrasing - I more thought it didn't make sense to have such a simple function if we could just do it on the datatype, but I guess we do that for `vars` etc. :P